### PR TITLE
Grid localization

### PR DESCRIPTION
**Pull request by amiller27**
Adds x-y localization using the bottom camera.  Implementation is complete (in that the overall structure and ideas probably won't change), but there are still bugs to be fixed before this is merged.

### CODE REVIEW

File: include/iarc7_vision/GridLineEstimator.hpp [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    /// pl_normal alone
+    void getPlanesForImageLines(const std::vector<cv::Vec2f>& image_lines,
+                                const ros::Time& time,
+                                double focal_length,
+                                std::vector<Eigen::Vector3d>& pl_normals) const;
+
+    double getThetaForPlanes(const std::vector<Eigen::Vector3d>& pl_normals) const;
+
+    /// Returns signed distance of each line from the origin without accounting
+    /// for altitude.  To the actual signed distance, multiply by the distance
```
Comment (aftersomemath) [inline include/iarc7_vision/GridLineEstimator.hpp]: Typo, should be "to get"

File: include/iarc7_vision/GridLineEstimator.hpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    /// @param[in] position_estimate Vector from origin of map to origin of pl_normal frame ("level_quad" on the ground)
+    void get2dPosition(const std::vector<double>& para_signed_dists,
+                       const std::vector<double>& perp_signed_dists,
+                       double theta,
+                       double height_estimate,
+                       const Eigen::Vector2d& position_estimate,
+                       Eigen::Vector2d& position,
+                       Eigen::Matrix2d& covariance) const;
+
+    double gridLoss(const std::vector<double>& wrapped_dists,
```
Comment (aftersomemath) [inline include/iarc7_vision/GridLineEstimator.hpp]: Comment for what this is for?

File: include/iarc7_vision/GridLineEstimator.hpp [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+                    double dist) const;
+
+    void processImage(const cv::Mat& image, const ros::Time& time) const;
```
Comment (aftersomemath) [inline include/iarc7_vision/GridLineEstimator.hpp]: Comment for what this is for?

File: include/iarc7_vision/GridLineEstimator.hpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    /// forward vector into the ground plane and use distances from that point
+    /// instead of from the origin.
+    static void getUnAltifiedDistancesFromLines(
+        double theta,
+        const std::vector<Eigen::Vector3d>& para_line_normals,
+        const std::vector<Eigen::Vector3d>& perp_line_normals,
+        std::vector<double>& para_signed_dists,
+        std::vector<double>& perp_signed_dists);
+
+    void get1dGridShift(const std::vector<double>& wrapped_dists,
```
Comment (aftersomemath) [inline include/iarc7_vision/GridLineEstimator.hpp]: Comment for what this is for?

File: src/GridLineEstimator.cpp [hunk 5]
```diff
@@ -0,0 +1,870 @@
...
+
+void GridLineEstimator::processImage(const cv::Mat& image,
+                                     const ros::Time& time) const
+{
+    const double height = last_filtered_position_(2);
+
+    // Extract lines from image
+    std::vector<cv::Vec2f> lines;
+    getLines(lines, image, height);
+    ROS_INFO("Number of lines extracted: %lu", lines.size());
```
Comment (aftersomemath) [inline src/GridLineEstimator.cpp]: ROS_DEBUG?

File: src/GridLineEstimator.cpp [hunk 6]
```diff
@@ -0,0 +1,870 @@
...
+        p.x = -10;
+        p.y = (pl_normal(2)*height - pl_normal(0)*p.x) / pl_normal(1);
+        lines_marker.points.push_back(p);
+    }
+    lines_marker.color.a = 1;
+    lines_marker.color.r = 0;
+    lines_marker.color.g = 1;
+    lines_marker.color.b = 1;
+    lines_marker.scale.x = 0.02;
+    debug_line_markers_pub_.publish(lines_marker);
```
Comment (aftersomemath) [inline src/GridLineEstimator.cpp]: Using publisher that might not be initialized?

File: src/GridLineEstimator.cpp [hunk 7]
```diff
@@ -0,0 +1,870 @@
...
+    p.x = std::cos(yaw);
+    p.y = -std::sin(yaw);
+    direction_marker.points.push_back(p);
+    direction_marker.color.a = 1;
+    direction_marker.color.r = 0;
+    direction_marker.color.g = 1;
+    direction_marker.color.b = 1;
+    direction_marker.scale.x = 0.05;
+    direction_marker.scale.y = 0.08;
+    debug_direction_marker_pub_.publish(direction_marker);
```
Comment (aftersomemath) [inline src/GridLineEstimator.cpp]: Using publisher that might not be initialized?

File: src/GridLineEstimator.cpp [hunk 8]
```diff
@@ -0,0 +1,870 @@
...
+                  perp_signed_dists,
+                  best_theta,
+                  height,
+                  last_position_2d,
+                  position_2d,
+                  position_cov_2d);
+
+    // Publish updated position
+    geometry_msgs::TransformStamped camera_to_lq_transform;
+    ROS_ASSERT(transform_wrapper_.getTransformAtTime(camera_to_lq_transform,
```
Comment (aftersomemath) [inline src/GridLineEstimator.cpp:799]: I'm not sure we want to handle this like we do in LLM. Especially due to the transform errors we see occasionally.

File: src/GridLineEstimator.cpp [hunk 9]
```diff
@@ -0,0 +1,870 @@
...
+
+    pose_msg.pose.pose.position.x = position(0);
+    pose_msg.pose.pose.position.y = position(1);
+    pose_msg.pose.pose.position.z = position(2);
+
+    ROS_ASSERT_MSG(std::isfinite(position(0)), "%f", position(0));
+    ROS_ASSERT_MSG(std::isfinite(position(1)), "%f", position(1));
+    ROS_ASSERT_MSG(std::isfinite(position(2)), "%f", position(2));
+
+    pose_msg.pose.pose.orientation.w = 1;
```
Comment (aftersomemath) [inline src/GridLineEstimator.cpp:841]: Comment that these might get filled in later by the yaw estimate?

File: src/GridLineEstimator.cpp (hunk 8 above)
Comment (aftersomemath) [reply src/GridLineEstimator.cpp:799]: As discussed let's fail the update and try again.

File: src/GridLineEstimator.cpp (hunk 6 above)
Comment (amiller27) [reply src/GridLineEstimator.cpp]: Fixed

File: src/GridLineEstimator.cpp (hunk 7 above)
Comment (amiller27) [reply src/GridLineEstimator.cpp]: Fixed

File: src/GridLineEstimator.cpp (hunk 5 above)
Comment (amiller27) [reply src/GridLineEstimator.cpp]: Fixed

File: src/GridLineEstimator.cpp (hunk 8 above)
Comment (amiller27) [reply src/GridLineEstimator.cpp:799]: Fixed

File: include/iarc7_vision/GridLineEstimator.hpp (hunk 1 above)
Comment (amiller27) [reply include/iarc7_vision/GridLineEstimator.hpp]: Fixed

File: include/iarc7_vision/GridLineEstimator.hpp (hunk 4 above)
Comment (amiller27) [reply include/iarc7_vision/GridLineEstimator.hpp]: Fixed

File: include/iarc7_vision/GridLineEstimator.hpp (hunk 2 above)
Comment (amiller27) [reply include/iarc7_vision/GridLineEstimator.hpp]: Fixed

File: include/iarc7_vision/GridLineEstimator.hpp (hunk 3 above)
Comment (amiller27) [reply include/iarc7_vision/GridLineEstimator.hpp]: Fixed

**Review by aftersomemath — APPROVED** [review]

**Pull request merged**